File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        List<Integer> cur = new ArrayList();
        dfs(0,cur,ans, candidates, target);
        return ans;
        
    }
    private void dfs(int index, List<Integer> cur, List<List<Integer>> ans, int[] candidates, int target){
        if(target ==0){
            ans.add(new ArrayList(cur));
        } else if (target < 0 || index >= candidates.length){
            return;
        }else{
            cur.add(candidates[index]);
            dfs(index,cur,ans,candidates,target-candidates[index]);

            cur.remove(cur.get(cur.size()-1));
            dfs(index+1,cur,ans,candidates,target);
        }
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res,sol=[],[]
        nums = candidates
        n= len(nums)

        def dfs(i, cur_summ):
            if cur_summ == target:
                res.append(sol[:])
                return
            
            if cur_summ > target or i==n:
                return

            dfs(i+1,cur_summ)
            sol.append(nums[i])
            dfs(i,cur_summ+nums[i])
            sol.remove(nums[i])
        dfs(0,0)
        return res




















        # res = []

        # def dfs(i, cur, total):
        #     if total == target:
        #         res.append(cur[:])
        #         return
            
        #     if i >= len(candidates) or total > target:
        #         return

        #     cur.append(candidates[i])
        #     dfs(i, cur, total + candidates[i])
        #     cur.pop()
        #     dfs(i + 1, cur, total)

        # dfs(0, [], 0)
        # return res
        